File: 383-ransom-note/383-ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> ransom;
        unordered_map<char, int> mag;
        
        for (char c : ransomNote) {
            ransom[c]++;
        }
        
        for (char c : magazine) {
            mag[c]++;
        }
        
        for (const auto key : ransom) {
            if (!mag[key.first] || mag[key.first] < key.second) {
                return false;
            } 
        }
        
        return true;
    }
};

File: 383.ransom-note.py
# @leet start
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransomCount = Counter(ransomNote)
        magazineCount = Counter(magazine)

        for letter in ransomCount:
            if letter not in magazineCount:
                return False
            if ransomCount[letter] > magazineCount[letter]:
                return False

        return True
# @leet end
